### PR TITLE
Add hardlink support for Linux & macOS

### PR DESCRIPTION
**Pull request by BlauFx**
https://user-images.githubusercontent.com/34138240/209813281-0548784f-35db-4349-be90-56caf5f1b3b1.mp4

### CODE REVIEW

File: osu.Game/IO/HardLinkHelper.cs
Comment (bdach) [inline osu.Game/IO/HardLinkHelper.cs:38]: Can this become its own method now that it has multiple platform implementations? Kinda feels bad seeing this same platform choice code in 2 places (here and `RealmFileStore`).

File: osu.Game/IO/HardLinkHelper.cs
Comment (bdach) [inline osu.Game/IO/HardLinkHelper.cs:79]:
```suggestion
        [DllImport("libc", SetLastError = true)]
```

**Comment by bdach** [comment]
Also this message on the first run setup overlay is a bit wrong now:

https://github.com/ppy/osu/blob/e4c060fc3afa1991fc1e05b79f9134c10713f306/osu.Game/Overlays/FirstRunSetup/ScreenImportFromStable.cs#L139-L140

**Comment by bdach** [comment]
I started going through this pull again but ended up pretty much rewriting it for multiple reasons elaborated in commit messages. I don't think this was actually tested either because there was an egregious logic bug that meant that this not only didn't work on linux, but also broke windows (49b0ec9ddb43d0f507ba8155d7bec69ed7876bb8).

Going to need a second review from someone. And going to need to re-test hard links on windows too to make sure nothing is broken.

(As an aside I do wonder if macOS can just use the same linux implementation...)

**Comment by peppy** [comment]
This looks to work fine on macOS so I've added support.

**Review by peppy — DISMISSED** [review]

File: osu.Game/Overlays/FirstRunSetup/ScreenImportFromStable.cs
Comment (bdach) [inline osu.Game/Overlays/FirstRunSetup/ScreenImportFromStable.cs:135]: @peppy I think this is going to need updating to `IsDesktop` too? Would entail trying to migrate across filesystems.

**Review by peppy — APPROVED** [review]

**Review by bdach — APPROVED** [review]
Re-tested windows. Looks fine to me.

**Pull request merged**
